Fix that collection associations were being cached with the current_scope information

Empact opened the pull request:
This would mean that `.scoping` values would be retained in the association proxy
cache even after exiting the scoping block, because certain modifications would
not be cleared by `#reset` - for example adding modules via `.extending`.

Now the proxy cache maintains a hash where the proxies are cached based on
`owner#new_record?` status and the `klass`'s `current_scope#object_id`. If
there is no active `current_scope`, the default scope will be active - cached
under `nil#object_id`.

An alternative to `#object_id` would have been to use `#to_sql`, which is used
e.g. by `Relation#==`, this would be somewhat more explicit but would incur
extra cost in evaluating the relation at `#reader` time rather than when data
is actually read from the database.

Empact commented:
There might be a more elegant way to do this... e.g. keying on `klass.all.extending_values`. I'm not necessarily familiar with all the options / concerns here.

Empact commented:
I prefer #19375 as a solution to this.

sgrif commented:
Can you clarify what method you're referring to when you say `#reset`?

sgrif commented on activerecord/lib/active_record/associations/collection_association.rb:
```diff
@@ -38,7 +38,10 @@ def reader(force_reload = false)
+          # Include the current_scope#object_id so that the association will
+          # not retain .scoping information after leaving a .scoping context.
+          @proxy ||= {}
+          @proxy[klass.current_scope.object_id] ||= CollectionProxy.create(klass, self)
```
This entirely defeats the purpose of the cache, as this value changes very frequently.

Empact commented:
`Relation#reset`

sgrif commented:
Why would you expect that to remove the effects of `.none`? `.reset` clears any caches so the query gets re-executed, but it doesn't change the query.

Empact commented:
Could you take a look at the test included in the PR? It illustrates the issue.

sgrif commented:
Yes, I see the issue illustrated by the test. I'm trying to get a better understanding of where the leak is occurring and why you chose this implementation.

Empact commented:
AFAIK what happens is the default_scope is `none!`ed, and then subsequently fetched in a noned state, even after existing the scoping block.

sgrif commented:
This demonstrates the flaw in this solution:

```ruby
a_relation_which_should_not_change_later = @welcome.comments

Comment.none.scoping do
  assert a_relation_which_should_not_change_later.any?
end
```

sgrif commented:
Actually I'm starting to wonder why associations are respecting the current scope at all. I'm sure there's a reason, but it seems unclear

sgrif commented:
> AFAIK what happens is the default_scope is none!ed, and then subsequently fetched in a noned state, even after existing the scoping block.

Not sure where you're getting that from. We never mutate the default scope. It looks like what's happening is the association instance is created inside of a `scoping` block, (and read inside of that block), which causes the collection proxy instance to get created with that scope. Since the same association instance is used in both blocks, the same collection proxy exists.

It seems like the association should be ignoring the current scope entirely to me. /cc @rafaelfranca @matthewd

Empact commented:
> We never mutate the default scope.

`Relation.all` returns a noned `#relation` from `#build_default_scoped`, which is then stored as the `target_scope` in the `CollectionProxy` - this is merged with the `association_scope` on read. Again I'm not 100% clear on the paths taken through the code but this is how it begins AFAIK.

sgrif commented:
Right, which again doesn't mutate anything. `.all` returns a new object every time you call it.

Empact commented:
I'll take another look at it tonight, try to pin it down exactly.

Empact committed:
Don't extend CollectionProxy with klass.current_scope

This introduces `proxy_scope` as an alternative to `scope` for
merging into CollectionProxy. The only difference is that it merges
in `default_scoped` rather than `all`, so that it never picks up
and caches values from `current_scope`.

`current_scope` continues to be picked up via existing calles to
`scope` and `target_scope`.

I also switched from `merge` to `merge!` in `target_scope` as that
saves a `spawn` call, which is unnecessary as the
`association_relation` is newly created for each call.

Empact commented:
@sgrif I took another look tonight and I think I have a proper fix - as you suggested I dropped `current_scope` from the initial `CollectionProxy#merge!`, using `default_scoped` instead, so only that much is cached. What do you think?
